dataworkspace/dataworkspace/apps/applications/gitlab.py

```diff
@@ -47,3 +48,60 @@ def gitlab_api_v4_ecr_pipeline_trigger(
     ).json()
     logger.debug('Started pipeline: %s', pipeline)
     return pipeline
+
+
+def gitlab_has_developer_access(user, gitlab_project_id):
+    # Having developer access to a project is cached to mitigate slow requests
+    # to GitLab. _Not_ having developer access to not cached to allow granting
+    # of access to have an immediate effect
+    #
+    # This would still mean that once expired, the user would still have a
+    # slower request. This could be avoided by a background job that polls
+    # for permissions changes before they expire. Leaving that until there is
+    # more evidence that a) the semantic behaviour is indeed what we want, and
+    # b) the performance is slow enough to justify that.
+    #
+    # The caching is suspected to be particularly important for visualisation
+    # previews: they can have multiple HTTP requests that each need to be
+    # checked for authorisation. This includes websocket requests that are
+    # expected to behave in almost real time. Websocket connections often drop
+    # out, so even once the visualisation is loaded, a reconnection, which
+    # would then need another authorisation check, should be speedy.
+    cache_key = f'gitlab-developer--{gitlab_project_id}--{user.id}'
+    has_access = cache.get(cache_key)
+    if has_access:
+        return True
+
+    gitlab_users, status = gitlab_api_v4_with_status(
+        'GET',
+        f'/users',
+        params=(('extern_uid', user.profile.sso_id), ('provider', 'oauth2_generic')),
+    )
+
+    if status != 200:
+        raise Exception(
+            f'Unable to find GitLab user for {user.profile.sso_id}: received {status}'
+        )
+
+    if len(gitlab_users) > 1:
+        raise Exception(f'Too many GitLab users matching {user.profile.sso_id}')
+
+    if not gitlab_users:
+        return False
+
+    gitlab_user = gitlab_users[0]
+    gitlab_project_users = gitlab_api_v4(
+        'GET',
+        f'/projects/{gitlab_project_id}/members/all',
+        params=(('user_ids', str(gitlab_user['id'])),),
+    )
+
+    has_access = True in (
+        gitlab_project_user['id'] == gitlab_user['id']
+        and gitlab_project_user['access_level'] >= int(DEVELOPER_ACCESS_LEVEL)
+        for gitlab_project_user in gitlab_project_users
```

Comment:
```suggestion
    has_access = any(
        gitlab_project_user['id'] == gitlab_user['id']
        and gitlab_project_user['access_level'] >= int(DEVELOPER_ACCESS_LEVEL)
        for gitlab_project_user in gitlab_project_users
    )
```

Functionally-equivalent, but believe this is _slightly_ more "Pythonic". Take it or leave it obviously.

Reply:
Agreed: will change